waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingServiceTest.java

```diff
@@ -355,4 +360,118 @@ public void panopticOperationsHandlerSetUgi() throws Exception {
     assertThat(result, containsInAnyOrder("ugi", "ugi2"));
   }
 
+  @Test(expected = NoPrimaryMetastoreException.class)
+  public void noPrimaryMappingThrowsException() {
+    when(metaStoreMappingFactory.newInstance(federatedMetastore)).thenReturn(metaStoreMappingFederated);
+    AbstractMetaStore unavailableMetastore = newFederatedInstance("name2", "thrift:host:port");
+    MetaStoreMapping unavailableMapping = mockNewMapping(false, "name2_");
+    when(metaStoreMappingFactory.newInstance(unavailableMetastore)).thenReturn(unavailableMapping);
+
+    service = new PrefixBasedDatabaseMappingService(metaStoreMappingFactory,
+        Arrays.asList(federatedMetastore, unavailableMetastore), queryMapping);
+    service.primaryDatabaseMapping();
```

Comment:
I can't quite remember how this all works, what is it in the above code that causes the service to not find a primary metastore?

Reply:
When PrefixBasedDatabaseMappingService is constructed, if the list of metastore passed as parameters contains a primary metastore, a local variable, `primaryDatabaseMapping` is set to map to the primary, otherwise it is null. When `service.primaryDatabaseMapping()` is called, `primaryDatabaseMapping` is checked, and if it's null, an exception is thrown.
In this test, the prefix service doesn't have a primary passed through.